### DISCUSSION TITLE
Constraint usage

### DISCUSSION THREAD

**Discussion by sstroemer** [calliope-project/calliope#479] [General]
What is the design reason to handle stuff like `cost(monetary__region1_1__csp_)` as a constraint? I know that expression usage in pyomo is expensive, and that it makes sense to keep "cost expressions" for easy postprocessing / extraction of cost components, but do we have a good idea of the involved performance gain/loss of constructing constraints instead of expressions?

**Comment by brynpickering** [comment]
I moved to using expressions for costs in a different branch I was working on: https://github.com/calliope-project/calliope/blob/euro-calliope/calliope/backend/pyomo/constraints/costs.py

We're also using expressions for costs in v0.7. 

I'm not sure exactly what the performance hit is. I seem to remember that there wasn't a hit for solving the optimisation problem - e.g., gurobi seems to handle these cases quite well - but the pyomo model had a lower memory footprint with expressions rather than variables (because you don't need a variable AND and constraint to set that variable's value).
